File: pool-cpu/lib/hobc_marker.js
'use strict';
/*
 * HOBC V6 coinbase telemetry marker builder (KPSS GPU pool, KawPow).
 *
 * Wire format matches the node consensus parser and SHA ckpool builder:
 *   magic "HOBC" (4) + version 1 (1) + canonical-order TLVs:
 *     0x01 pool_id            len 1   REQUIRED
 *     0x02 algo               len 1   REQUIRED (0=sha256 1=kawpow 2=randomx)
 *     0x03 winning_share_diff len 4   REQUIRED (IEEE-754 float32, little-endian)
 *     0x04 pool_site          len<=24 optional (consensus max; last chars kept)
 *     0x06 worker_name        len 16  optional but always emitted (space-padded)
 *
 * Rig (0x05) is omitted so site/worker fit under the 100-byte scriptSig cap with
 * headroom. Long text fields keep the last N characters.
 */

const SITE_MAX = 24;
const WORKER_LEN = 16;

// difficulty-1 target as used for compact-bits difficulty (0x1d00ffff)
const DIFF1 = 0xffff * Math.pow(2, 8 * (0x1d - 3));

function bitsToDifficulty(bitsHex) {
  const bits = parseInt(bitsHex, 16) >>> 0;
  const exponent = bits >>> 24;
  const mantissa = bits & 0x007fffff;
  const target = mantissa * Math.pow(2, 8 * (exponent - 3));
  if (!isFinite(target) || target <= 0) return 0;
  return DIFF1 / target;
}

/** Space-pad to len; keep the last len chars when src is longer. */
function fillField(len, src) {
  const out = Buffer.alloc(len, 0x20);
  if (!src) return out;
  let s = '';
  const raw = String(src);
  for (let i = 0; i < raw.length; i++) {
    const c = raw.charCodeAt(i);
    if (c >= 0x20 && c <= 0x7e) s += raw[i];
  }
  if (s.length > len) s = s.slice(s.length - len);
  for (let i = 0; i < s.length; i++) out[i] = s.charCodeAt(i);
  return out;
}

function workerLabel(workerName) {
  if (!workerName) return '';
  const s = String(workerName);
  const dot = s.lastIndexOf('.');
  return (dot >= 0 && dot + 1 < s.length) ? s.slice(dot + 1) : s;
}

function rigLabel(userAgent) {
  if (!userAgent) return '';
  const s = String(userAgent);
  let out = '';
  for (let i = 0; i < s.length; i++) {
    const c = s.charCodeAt(i);
    if (c < 0x20 || c > 0x7e) continue;
    if (s[i] === '/' || s[i] === ' ') break;
    out += s[i];
    if (out.length >= 64) break;
  }
  return out;
}

function buildMarker(poolId, algo, winningShareDiff, poolSite, rigModel, workerName) {
  const parts = [];
  parts.push(Buffer.from('HOBC', 'ascii'));
  parts.push(Buffer.from([1]));

  parts.push(Buffer.from([0x01, 1, poolId & 0xff]));
  parts.push(Buffer.from([0x02, 1, algo & 0xff]));

  const diffBuf = Buffer.alloc(6);
  diffBuf[0] = 0x03; diffBuf[1] = 4;
  diffBuf.writeFloatLE(Math.fround(winningShareDiff || 0), 2);
  parts.push(diffBuf);

  if (poolSite && String(poolSite).length) {
    const site = fillField(SITE_MAX, poolSite);
    let sl = SITE_MAX;
    while (sl > 0 && site[sl - 1] === 0x20) sl--;
    if (sl > 0) {
      parts.push(Buffer.concat([Buffer.from([0x04, sl]), site.slice(0, sl)]));
    }
  }

  // 0x05 rig omitted — budget reserved for site/worker + scriptSig headroom
  void rigModel;

  const worker = fillField(WORKER_LEN, workerName ? workerLabel(workerName) : '');
  parts.push(Buffer.concat([Buffer.from([0x06, WORKER_LEN]), worker]));

  return Buffer.concat(parts);
}

function patchMarkerIdentity(marker, rigModel, workerName) {
  if (!Buffer.isBuffer(marker) || marker.length < 5) return 0;
  if (marker.toString('ascii', 0, 4) !== 'HOBC' || marker[4] !== 1) return 0;
  void rigModel;
  const worker = fillField(WORKER_LEN, workerName);
  let i = 5;
  let patched = 0;
  while (i + 2 <= marker.length) {
    const t = marker[i];
    const l = marker[i + 1];
    if (i + 2 + l > marker.length) break;
    if (t === 0x06 && l === WORKER_LEN) {
      worker.copy(marker, i + 2);
      patched++;
    }
    i += 2 + l;
  }
  return patched;
}

function patchTxIdentity(txHex, rigModel, workerName) {
  const buf = Buffer.from(String(txHex), 'hex');
  const magic = Buffer.from('HOBC', 'ascii');
  const idx = buf.indexOf(magic);
  if (idx < 0 || idx + 5 > buf.length || buf[idx + 4] !== 1) return null;
  const marker = buf.slice(idx);
  if (patchMarkerIdentity(marker, rigModel, workerName) <= 0) return null;
  return buf.toString('hex');
}

module.exports = {
  buildMarker,
  bitsToDifficulty,
  workerLabel,
  rigLabel,
  patchMarkerIdentity,
  patchTxIdentity,
  WORKER_LEN,
  SITE_MAX,
};

File: pool-sha/hobc-pplns/src/hobc_marker.c
/*
 * HOBC V6 coinbase telemetry marker builder (pool side). See hobc_marker.h.
 */
#include "hobc_marker.h"

#include <ctype.h>
#include <string.h>

void hobc_marker_fill_field(char *dst, size_t dstlen, const char *src)
{
	size_t i, len;
	const char *use;

	if (!dst || dstlen == 0)
		return;
	memset(dst, ' ', dstlen);
	if (!src || !src[0])
		return;
	/* src must be a NUL-terminated C string. Keep the last dstlen chars when longer. */
	len = strlen(src);
	use = src;
	if (len > dstlen) {
		use = src + (len - dstlen);
		len = dstlen;
	}
	for (i = 0; i < len; i++) {
		unsigned char c = (unsigned char)use[i];
		dst[i] = (c >= 0x20 && c <= 0x7e) ? (char)c : ' ';
	}
}

void hobc_marker_worker_label(char *dst, size_t dstlen, const char *workername)
{
	const char *src = workername;
	const char *dot;

	if (!dst || dstlen == 0)
		return;
	if (!workername || !workername[0]) {
		hobc_marker_fill_field(dst, dstlen, "");
		return;
	}
	/* Prefer the last dotted suffix (rig / worker id), not the payout address. */
	dot = strrchr(workername, '.');
	if (dot && dot[1])
		src = dot + 1;
	/* fill_field keeps the last dstlen chars when the suffix is still too long. */
	hobc_marker_fill_field(dst, dstlen, src);
}

void hobc_marker_rig_label(char *dst, size_t dstlen, const char *useragent)
{
	char tmp[64];
	size_t i, n = 0;

	if (!dst || dstlen == 0)
		return;
	if (!useragent || !useragent[0]) {
		hobc_marker_fill_field(dst, dstlen, "");
		return;
	}
	for (i = 0; useragent[i] && n < sizeof(tmp) - 1; i++) {
		unsigned char c = (unsigned char)useragent[i];
		if (c < 0x20 || c > 0x7e)
			continue;
		if (c == '/' || c == ' ')
			break;
		tmp[n++] = (char)c;
	}
	tmp[n] = '\0';
	hobc_marker_fill_field(dst, dstlen, tmp);
}

int hobc_build_marker(uint8_t *out, int pool_id, int algo,
		      float winning_share_diff, const char *pool_site,
		      const char *rig_model, const char *worker_name)
{
	char site[HOBC_MARKER_SITE_MAX];
	char worker[HOBC_MARKER_WORKER_LEN];
	int n = 0;
	size_t sl;

	(void)rig_model; /* 0x05 omitted — budget goes to site/worker + 1-byte headroom */

	if (!out)
		return 0;

	hobc_marker_fill_field(site, sizeof(site), pool_site);
	sl = HOBC_MARKER_SITE_MAX;
	while (sl > 0 && site[sl - 1] == ' ')
		sl--;

	/* magic + version */
	out[n++] = 'H';
	out[n++] = 'O';
	out[n++] = 'B';
	out[n++] = 'C';
	out[n++] = 1;

	/* 0x01 pool_id (len 1) */
	out[n++] = 0x01;
	out[n++] = 1;
	out[n++] = (uint8_t)(pool_id & 0xff);

	/* 0x02 algo (len 1) */
	out[n++] = 0x02;
	out[n++] = 1;
	out[n++] = (uint8_t)(algo & 0xff);

	/* 0x03 winning_share_diff (len 4, float32 little-endian) */
	out[n++] = 0x03;
	out[n++] = 4;
	memcpy(out + n, &winning_share_diff, 4);
	n += 4;

	/* 0x04 pool_site (optional; omit when empty) */
	if (sl > 0) {
		out[n++] = 0x04;
		out[n++] = (uint8_t)sl;
		memcpy(out + n, site, sl);
		n += (int)sl;
	}

	/* 0x06 worker_name — always fixed length (space-padded) for in-place patch */
	if (worker_name && worker_name[0])
		hobc_marker_worker_label(worker, sizeof(worker), worker_name);
	else
		hobc_marker_fill_field(worker, sizeof(worker), "");
	out[n++] = 0x06;
	out[n++] = (uint8_t)HOBC_MARKER_WORKER_LEN;
	memcpy(out + n, worker, HOBC_MARKER_WORKER_LEN);
	n += HOBC_MARKER_WORKER_LEN;

	return n;
}

int hobc_marker_patch_identity(uint8_t *marker, int mlen,
			       const char *rig_model, const char *worker_name)
{
	char worker[HOBC_MARKER_WORKER_LEN];
	int i = 5;
	int patched = 0;

	(void)rig_model;

	if (!marker || mlen < 5)
		return 0;
	if (marker[0] != 'H' || marker[1] != 'O' || marker[2] != 'B' ||
	    marker[3] != 'C' || marker[4] != 1)
		return 0;

	/* Callers pass already-labeled, space-padded fixed worker field (no NUL). */
	memset(worker, ' ', sizeof(worker));
	if (worker_name)
		memcpy(worker, worker_name, HOBC_MARKER_WORKER_LEN);

	while (i + 2 <= mlen) {
		uint8_t t = marker[i];
		uint8_t l = marker[i + 1];

		if (i + 2 + l > mlen)
			break;
		if (t == 0x06 && l == HOBC_MARKER_WORKER_LEN) {
			memcpy(marker + i + 2, worker, HOBC_MARKER_WORKER_LEN);
			patched++;
		}
		i += 2 + l;
	}
	return patched;
}

int hobc_coinb2_patch_identity(uint8_t *coinb2, int coinb2len,
			       const char *rig_model, const char *worker_name)
{
	int i;

	if (!coinb2 || coinb2len < 5)
		return 0;
	for (i = 0; i <= coinb2len - 5; i++) {
		if (coinb2[i] == 'H' && coinb2[i + 1] == 'O' &&
		    coinb2[i + 2] == 'B' && coinb2[i + 3] == 'C' &&
		    coinb2[i + 4] == 1)
			return hobc_marker_patch_identity(coinb2 + i, coinb2len - i,
							  rig_model, worker_name);
	}
	return 0;
}
